Improvement of CCS usage stats collection

smalyshev opened the pull request:
Implements collecting CCS usage stats and generating stats snapshots

Features covered:
- [x] Collect per-request statistics in CCSUsage
- [x] Collect overall statistics in CCSUsageTelemetry
- [x] Produce snapshots as CCSTelemetrySnapshot
- [x] Generate XContent from the snapshot
- [x] Client recognition
- [ ] Add logic to collect and merge telemetry snapshots for stats action processing
- [ ] Add metrics to the cluster stats output
- [ ] Test that all paths through the search collect proper metrics

Also in the patch: 
- Create `LongMetric` class that encapsulates metric handling (mostly for latencies) and `LongMetricValue` as a snapshot
- Histograms will be part of the on-wire message, since they seem to require only ~2K size with 2 significant digits. 
- `SearchResponseActionListener` as a separate class which also takes care of metrics
- Metrics are built both on success and failure, though we could use some more fine grained failure recognition later
- TODO: think of how to get rid of `isAsyncSearchTask`.

smalyshev committed:
Refactoring of usage stats collection

Implements collecting CCS usage stats and generating stats snapshots

elasticsearchmachine commented:
Pinging @elastic/es-search-foundations (Team:Search Foundations)

smalyshev committed:
Add more tests and fix some test-detected bugs

smalyshev committed:
More test & fix snapshot merging

smalyshev committed:
Add more tests & comments

smalyshev committed:
Implement client extraction

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java:
```diff
@@ -8,172 +8,174 @@
+    public static final String MRT_FEATURE = "mrt_on";
+    public static final String ASYNC_FEATURE = "async";
+    public static final String WILDCARD_FEATURE = "wildcards";
+    // TODO: Do we need to count unknown clients separately?
```
probably not, as you say - just calculate it on the fly

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java:
```diff
@@ -8,172 +8,174 @@
 
     public void updateUsage(CCSUsage ccsUsage) {
+        // Ignore empty usage objects, we only care for true cross-cluster requests
+        if (ccsUsage.getRemotesCount() == 0) {
```
Should this ever happen? If no, then just add an assert here that remotes count > 0.

smalyshev replied:
It shouldn't. I think you're right, assert here is better, I'll change it.

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java:
```diff
@@ -8,172 +8,174 @@
+        long searchTook = ccsUsage.getTook();
+        if (isSuccess(ccsUsage)) {
+            successCount.increment();
+            // TODO: do we need to count latencies for failed requests? Right now the code doesn't collect it without SearchResponse
```
No, I don't think we need latencies for failure scenarios.

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java:
```diff
@@ -8,172 +8,174 @@
+                tookMrtFalse.record(searchTook);
             }
-
+            // TODO: can we do it on failure? Not sure what "took" values contain in that case if at all.
```
Sometimes error responses will have took values, other times not. My vote is just to not record latencies for failure scenarios.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -396,8 +359,25 @@ void executeRequest(
+                        tl.setClient(client);
+                    }
+                    // TODO: is this the right way to check for wildcards?
+                    if (Arrays.stream(original.indices()).anyMatch(Regex::isSimpleMatchPattern)) {
```
Take a look at how the SearchResponse gets created.

For the query:

```
POST blogs,*:*/_async_search?ccs_minimize_roundtrips=true
```

You can see in the response that it parsed out just the indices as requested by the user:

```json
      "details": {
        "(local)": {
          "status": "running",
          "indices": "blogs",
          "timed_out": false
        },
        "remote2": {
          "status": "running",
          "indices": "*",
          "timed_out": false
        },
        "remote1": {
          "status": "skipped",
          "indices": "*",
          "timed_out": false,
          "failures": [
```

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java:
```diff
@@ -184,9 +186,43 @@ public String toString() {
+
+    // TODO: I wonder if it wouldn't be more correct if this lived on the Snapshot side,
+    // but SearchUsage does it this way so following the pattern here.
+    public CCSTelemetrySnapshot getCCSTelemetrySnapshot() {
```
Does this method need to be `synchronized` (or least a subsection of this method)? Since we use locks to do the updates we may need to have a lock to grab the snapshot. On the other hand since (currently) we also have LongAdders in many places, perhaps we can avoid a lock here? That would need a careful analysis. 

One risk of not using locking here is if someone adds to the Telemetry object later and doesn't ensure it is thread safe on the snapshot side (e.g., using a long instead of a LongAdder).

smalyshev replied:
Good question. I didn't fully consider the concurrency aspect yet, but certainly adding `synchronized` would make it safe. I am not 100% sure it's required yet, since if we have read while we're updating worst that could happen is that we get slightly inconsistent data (i.e. total counter incremented, but some other counter isn't yet) or such, which is not a huge deal. But I need to look into it more carefully to be sure.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -1778,4 +1740,124 @@ List<SearchShardIterator> getLocalShardsIterator(
+         */
+        private void extractCCSTelemetry(SearchResponse searchResponse) {
+            usageBuilder.took(searchResponse.getTookInMillis());
+            // TODO: what happens with the local cluster there? Are we tracking it too just like the others?
```
Good question - I don't think we need/want to track the local cluster individually, but probably best to check with Product.

smalyshev replied:
For now I am assuming it would just produce `(local)` entry and I leave it as is, but I'll check it later and it would be easy to add a check here that removes the local one.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -1778,4 +1740,124 @@ List<SearchShardIterator> getLocalShardsIterator(
+            // TODO: what happens with the local cluster there? Are we tracking it too just like the others?
+            for (String clusterAlias : searchResponse.getClusters().getClusterAliases()) {
+                SearchResponse.Cluster cluster = searchResponse.getClusters().getCluster(clusterAlias);
+                // TODO: check if getTook is meaningful if cluster was skipped
```
A cluster with `skipped` status will not have a took time. You can check the toXContent rendering of the SearchResponse.Clusters.

Example:

```json
        "remote1": {
          "status": "skipped",
          "indices": "*",
          "timed_out": false,
          "failures": [
            {
              "shard": -1,
              "index": null,
              "reason": {
                "type": "search_phase_execution_exception",
                "reason": "all shards failed",
                "phase": "query",
                "grouped": true,
                "failed_shards": [
                  {
                    "shard": 0,
                    "index": "remote1:blogs",
                    "node": "wACt7av_RI26NO-SIf3oqA",
...
```

smalyshev replied:
I see, then the `getTook` part needs to be moved to the `else` clause.

quux00 reviewed (commented):
Good progress!
First round feedback left as individual comments.

smalyshev committed:
Incorporate feedback, resolve TODOs:

- Wildcard check uses resolvedIndices
- No latency on failed searches
- Not counting unknown clients
- updateUsage check is assert now

smalyshev committed:
Fix unknown clients test

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSTelemetrySnapshot.java:
```diff
@@ -0,0 +1,376 @@
+        return took;
+    }
+
+    public LongMetricValue getTookMrtTrue() {
```
Somewhere in this class can you put a code comment as to what "mrt" means. People not familiar with CCS will not know what that refers to.

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSTelemetrySnapshot.java:
```diff
@@ -0,0 +1,376 @@
+    public static class PerClusterCCSTelemetry implements Writeable, ToXContentFragment {
+        private long count;
+        private long skippedCount;
+        private final LongMetricValue took;
```
Same comment here as above - replace with 3 longs, max, avg, p90.

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSTelemetrySnapshot.java:
```diff
@@ -0,0 +1,376 @@
+                byRemoteCluster.forEach((name, clusterData) -> {
+                    try {
+                        builder.field(name, clusterData);
+                    } catch (IOException e) {
```
Asking to learn - why do we need this catch block? And if it gets an IOException, what is the consequence of throwing a RuntimeException? We don't want to crash a server over this issue.

smalyshev replied:
Because it's a lambda, and lambdas in Java can't have checked exceptions. So if it's not caught, Java complains about lambda missing checked exception (which `IOException` is) so it has to be converted to runtime. Lambdas and checked exceptions don't play nice unfortunately. I could unroll it to for loop if it feels nicer (I like `forEach` better but it's just personal preference).

smalyshev replied:
The probability it will actually throw in this particular line is minimal, there's not much there to cause it, but I don't have much choice since `field` is declared with checked exception.

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsage.java:
```diff
@@ -8,104 +8,133 @@
 
-        public Builder numSkippedRemotes(int skippedRemotes) {
-            this.skippedRemotes = skippedRemotes;
+        public Builder skipRemote(String remote) {
```
Maybe call this "skippedRemote". To me that shows that this is recording some past event, rather than taking an action to "skip" it.

quux00 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java:
```diff
@@ -8,172 +8,166 @@
+        }
+    }
+
+    // Not enum because we won't mind other places adding their own features
```
I'm not sure what this comment means - can you clarify. Are you saying that more features could be added? If yes, why would using an enum not work?

smalyshev replied:
Because of something like async - i.e. say we add a module that does some functionality like some third kind of search, beyond regular and async, and we want to count it. If we make enum, we'd have to make change to this class each time it happens. But if it's just a string then we can make that module just call `setFeature` and this class doesn't need to know.

That said, we don't actually have (yet?) a good way for async or other modules to call `setFeature` - I'd like to make it happen and eliminate the complex async detection code, but I left it for the future for now. So if we wanted the enum right now and switch to a string later we could (I think serialization would work the same?) I just made it a string with an eye to eventually making the solution where async and others take care of marking themselves without telemetry class knowing.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -405,8 +368,32 @@ void executeRequest(
+                    if (isAsyncSearchTask(task)) {
+                        tl.setFeature(CCSUsageTelemetry.ASYNC_FEATURE);
+                    }
+                    String client = task.getHeader(Task.X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER);
```
Asking to learn - did we confirm with Kibana or the devtools team that this is the place they will set their client info?

smalyshev replied:
I checked the source code and the flag is set there. I can also verbally confirm with them too.

quux00 reviewed (commented):
More feedback.

smalyshev commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSTelemetrySnapshot.java:
```diff
@@ -0,0 +1,376 @@
+    /**
+     * Add the provided stats to the ones held by the current instance, effectively merging the two
+     */
+    public void add(CCSTelemetrySnapshot stats) {
```
It _looks_ like those adds will always be called in the single thread since the code in `TransportClusterStatsAction` collects node responses in a list and then passes it to final response ctor - so that's where the adds will happen, and it's single thread.

smalyshev committed:
Address pull feedback - more comments, light refactoring

smalyshev committed:
Fix test - random histogram should be not empty

quux00 reviewed (approved):
LGTM
